File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.py
class Solution(object):
    def preOrder(self, root, result):
        if root:
            result.append(root.val)
            self.preOrder(root.left, result)
            self.preOrder(root.right, result)
            
    def preorderTraversal(self, root):
        array = []
        self.preOrder(root, array)
        return array

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.cpp
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        if (!root) return res;

        stack<TreeNode*> stk;
        stk.push(root);

        while (!stk.empty()){
            int size = stk.size();
            
            for (int i = 0; i < size; i++){
                TreeNode* node = stk.top();
                stk.pop();
                res.push_back(node->val);

                if (node->right) stk.push(node->right);
                if (node->left) stk.push(node->left);
            }
        }

        return res;
    }
};